src/cargo/core/resolver/mod.rs

```diff
@@ -884,80 +884,205 @@ fn generalize_conflicting(
             // it will only help backjump further if it is older then the critical_age
             cx.is_active(*p).expect("parent not currently active!?") < backtrack_critical_age
         })
     {
-        for critical_parents_dep in critical_parents_deps.iter() {
+        'dep: for critical_parents_dep in critical_parents_deps.iter() {
             // A dep is equivalent to one of the things it can resolve to.
             // Thus, if all the things it can resolve to have already ben determined
             // to be conflicting, then we can just say that we conflict with the parent.
-            if let Some(others) = registry
+            let mut con = conflicting_activations.clone();
+            con.remove(&backtrack_critical_id);
+            con.insert(*critical_parent, backtrack_critical_reason.clone());
+
+            for other in registry
                 .query(critical_parents_dep)
                 .expect("an already used dep now error!?")
                 .iter()
-                .rev() // the last one to be tried is the least likely to be in the cache, so start with that.
-                .map(|other| {
-                    past_conflicting_activations
-                        .find(
-                            dep,
-                            &|id| {
-                                if id == other.summary.package_id() {
-                                    // we are imagining that we used other instead
-                                    Some(backtrack_critical_age)
-                                } else {
-                                    cx.is_active(id).filter(|&age|
-                                        // we only care about things that are older then critical_age
-                                        age < backtrack_critical_age)
-                                }
-                            },
-                            Some(other.summary.package_id()),
-                        )
-                        .map(|con| (other.summary.package_id(), con))
-                })
-                .collect::<Option<Vec<(PackageId, &ConflictMap)>>>()
+                .rev()
+            // the last one to be tried is the least likely to be in the cache, so start with that.
             {
-                let mut con = conflicting_activations.clone();
-                // It is always valid to combine previously inserted conflicts.
-                // A, B are both known bad states each that can never be activated.
-                // A + B is redundant but cant be activated, as if
-                // A + B is active then A is active and we know that is not ok.
-                for (_, other) in &others {
-                    con.extend(other.iter().map(|(&id, re)| (id, re.clone())));
+                if (our_activation_key.map_or(false, |our| {
+                    other.summary.package_id().as_activations_key() == our
+                }) || our_link.map_or(false, |_| other.summary.links() == our_link))
```

Comment:
Instead of extracting a single element above would it be perhaps more clear to change these to:

```rust
if our_activation_keys.len() == 1 && our_activation_keys.contains(&other.summary.package_id().as_activations_key()) {
    // ..
}
```

or something like that?